corehq/apps/locations/views.py

```diff
@@ -102,42 +100,121 @@ def page_context(self):
     @property
     def settings_context(self):
         return {
-            'loc_types': [self._get_loctype_info(l) for l in self.domain_object.location_types],
+            'loc_types': map(
+                self._get_loctype_info,
+                LocationType.objects.by_domain(self.domain)
+            )
         }
 
     def _get_loctype_info(self, loctype):
         return {
             'name': loctype.name,
             'code': loctype.code,
-            'allowed_parents': [p or None for p in loctype.allowed_parents],
+            'allowed_parents': [loctype.parent_type.name
+                                if loctype.parent_type else None],
             'administrative': loctype.administrative,
             'shares_cases': loctype.shares_cases,
             'view_descendants': loctype.view_descendants
         }
 
     def post(self, request, *args, **kwargs):
+        # This is really ugly, it needs a refactor.
+        # There are some changes coming to this UI, so I'm just gonna wait
+        # for that.  I intend to remove "code" from the UI altogether, and
+        # use the actual location_type pk to reference it, not the name.
         payload = json.loads(request.POST.get('json'))
+        sql_loc_types = {}
 
-        def mk_loctype(loctype):
-            loctype['allowed_parents'] = [p or '' for p in loctype['allowed_parents']]
-            cleaned_code = unicode_slug(loctype['code'])
-            if cleaned_code != loctype['code']:
-                err = _(
-                    'Location type code "{code}" is invalid. No spaces or special characters are allowed. '
-                    'It has been replaced with "{new_code}".'
-                )
-                messages.warning(request, err.format(code=loctype['code'], new_code=cleaned_code))
-                loctype['code'] = cleaned_code
-            return LocationType(**loctype)
-
-        #TODO add server-side input validation here (currently validated on client)
+        def mk_loctype(name, code, allowed_parents, administrative,
+                       shares_cases, view_descendants):
+            if allowed_parents and allowed_parents[0]:
+                parent = sql_loc_types[allowed_parents[0]]
```

Comment:
(this seems like it could be hiding bugs if you passed in more than one allowed parent - another possible motivator for changing the function api)